File: benchmarks/tsb/bench_to_date_input_fn.ts
/**
 * Benchmark: toDateInput — normalize various date input types to Date objects.
 * Outputs JSON: {"function": "to_date_input_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { toDateInput } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 50;

const strings = Array.from({ length: SIZE }, (_, i) => `2020-${String(1 + (i % 12)).padStart(2, "0")}-01`);
const timestamps = Array.from({ length: SIZE }, (_, i) => Date.now() + i * 86_400_000);
const dates = Array.from({ length: SIZE }, (_, i) => new Date(2020, i % 12, 1 + (i % 28)));

for (let i = 0; i < WARMUP; i++) {
  for (const s of strings.slice(0, 100)) toDateInput(s);
  for (const t of timestamps.slice(0, 100)) toDateInput(t);
  for (const d of dates.slice(0, 100)) toDateInput(d);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  for (const s of strings) toDateInput(s);
  for (const t of timestamps) toDateInput(t);
  for (const d of dates) toDateInput(d);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "to_date_input_fn",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_to_date_input_fn.py
"""Benchmark: toDateInput — normalize various date inputs using pandas Timestamp.
Mirrors tsb bench_to_date_input_fn.ts.
"""
import json, time
import pandas as pd

SIZE = 50_000
WARMUP = 5
ITERATIONS = 50

strings = [f"2020-{1 + (i % 12):02d}-01" for i in range(SIZE)]
timestamps = [int((pd.Timestamp("2020-01-01").timestamp() + i * 86400) * 1000) for i in range(SIZE)]
dates = [pd.Timestamp(2020, 1 + (i % 12), 1 + (i % 28)) for i in range(SIZE)]

for _ in range(WARMUP):
    for s in strings[:100]:
        pd.Timestamp(s)
    for t in timestamps[:100]:
        pd.Timestamp(t, unit="ms")
    for d in dates[:100]:
        pd.Timestamp(d)

start = time.perf_counter()
for _ in range(ITERATIONS):
    for s in strings:
        pd.Timestamp(s)
    for t in timestamps:
        pd.Timestamp(t, unit="ms")
    for d in dates:
        pd.Timestamp(d)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "to_date_input_fn",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
